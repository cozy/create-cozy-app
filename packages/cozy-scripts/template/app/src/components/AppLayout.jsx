/* global cozy */
import React from 'react'
import { Outlet } from 'react-router-dom'
import cx from 'classnames'

import { Layout, Main, Content } from 'cozy-ui/transpiled/react/Layout'
import Sprite from 'cozy-ui/transpiled/react/Icon/Sprite'
import Alerter from 'cozy-ui/transpiled/react/Alerter'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'
import MuiCozyTheme from 'cozy-ui/transpiled/react/MuiCozyTheme'
import BarTitle from 'cozy-ui/transpiled/react/BarTitle'
import { useClient } from 'cozy-client'

import Sidebar from 'src/components/Sidebar'

const AppLayout = () => {
  const { t } = useI18n()
  const client = useClient()
  const { isMobile } = useBreakpoints()
  const { BarCenter } = cozy.bar

  return (
    <Layout>
      {isMobile && (
        <BarCenter>
          <MuiCozyTheme>
            <BarTitle>{client.appMetadata.slug}</BarTitle>
          </MuiCozyTheme>
        </BarCenter>
      )}
      <Sidebar />
      <Main>
        <Content
          className={cx({
            'u-mh-half u-mv-1': isMobile,
            'u-mh-2 u-mv-1': !isMobile
          })}
        >
          <Outlet />
        </Content>
      </Main>
      <Alerter t={t} />
      <Sprite />
    </Layout>
  )
}

export default AppLayout
