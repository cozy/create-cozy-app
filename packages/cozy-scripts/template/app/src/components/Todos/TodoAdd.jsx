import React, { Component } from 'react'

import { withClient } from 'cozy-client'
import Input from 'cozy-ui/react/Input'
import Label from 'cozy-ui/react/Label'
import Button from 'cozy-ui/react/Button'

import { TODOS_DOCTYPE } from 'doctypes'

export class TodoAdd extends Component {
  constructor(props, context) {
    super(props, context)
    // initial component state
    this.state = {
      todoToAdd: '',
      isWorking: false
    }
  }

  // handle input value change
  handleChange = event => {
    this.setState({ todoToAdd: event.target.value })
  }

  // create the new todo
  handleSubmit = async () => {
    const { todoToAdd } = this.state
    const { client } = this.props
    // reset the input and display a spinner during the process
    this.setState({ todoToAdd: '' })

    try {
      this.setState({ isWorking: true })
      const newDoc = { _type: TODOS_DOCTYPE, name: todoToAdd }
      await client.save(newDoc)
    } finally {
      this.setState(() => ({ isWorking: false }))
    }
  }

  render() {
    const { todoToAdd, isWorking } = this.state
    return (
      <div>
        <h2 className="u-mt-2 u-mb-half">Add a new Todo</h2>
        <form onSubmit={this.handleSubmit}>
          <Label htmlFor="todo-add-input">Todo name</Label>
          <Input
            value={todoToAdd}
            onChange={this.handleChange}
            id="todo-add-input"
          />
          <Button
            onClick={this.submit}
            type="submit"
            busy={isWorking}
            label="add"
            size="large"
            extension="narrow"
          />
        </form>
      </div>
    )
  }
}

// get mutations from the client to use createDocument
export default withClient(TodoAdd)
