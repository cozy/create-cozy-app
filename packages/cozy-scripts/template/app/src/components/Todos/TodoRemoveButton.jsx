import React, { Component } from 'react'

import Button from 'cozy-ui/react/Button'
import Icon from 'cozy-ui/react/Icon'
import { withClient } from 'cozy-client'

export class TodoRemoveButton extends Component {
  constructor(props) {
    super(props)
    // initial component state
    this.state = { isWorking: false }
  }

  // delete the related todo
  removeTodo = async () => {
    const { client, todo } = this.props
    this.setState({ isWorking: true })

    try {
      await client.destroy(todo)
    } finally {
      this.setState({ isWorking: false })
    }
  }

  render() {
    const { isWorking } = this.state
    return (
      <Button
        className="todo-remove-button"
        theme="danger"
        iconOnly
        label="Delete"
        busy={isWorking}
        disabled={isWorking}
        onClick={this.removeTodo}
        extension="narrow"
      >
        {!isWorking ? <Icon icon="trash" /> : null}
      </Button>
    )
  }
}

// get mutations from the client to use deleteDocument
export default withClient(TodoRemoveButton)
