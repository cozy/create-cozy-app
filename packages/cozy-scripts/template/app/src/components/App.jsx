import React from 'react'
import { Route, Switch, Redirect, HashRouter } from 'react-router-dom'
import { hot } from 'react-hot-loader'

import { Layout, Main, Content } from 'cozy-ui/transpiled/react/Layout'
import { Sprite as IconSprite } from 'cozy-ui/transpiled/react/Icon'

import Sidebar from './Sidebar'
import Todos from './Todos'
import Hello1 from './HelloViews/Hello1'
import Hello2 from './HelloViews/Hello2'

const App = () => (
  <HashRouter>
    <Layout>
      <Sidebar />
      <Main>
        <Content className="app-content">
          <Switch>
            <Route path="/todos" component={Todos} />
            <Route path="/viewhello1" component={Hello1} />
            <Route path="/viewhello2" component={Hello2} />
            <Redirect from="/" to="/todos" />
            <Redirect from="*" to="/todos" />
          </Switch>
        </Content>
      </Main>
      <IconSprite />
    </Layout>
  </HashRouter>
)

/*
  Enable Hot Module Reload using `react-hot-loader` here
  We enable it here since App is the main root component
  No need to use it anywhere else, it sould work for all
  child components
*/
export default hot(module)(App)
