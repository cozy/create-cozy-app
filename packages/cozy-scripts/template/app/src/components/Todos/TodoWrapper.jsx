import React from 'react'

import { useQuery, isQueryLoading } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import { getAllTodos } from 'src/utils/queries'
import TodoAdd from 'src/components/Todos/TodoAdd/TodoAdd'
import TodoList from 'src/components/Todos/TodoList/TodoList'

export const TodoWrapper = () => {
  const { data, ...rest } = useQuery(
    getAllTodos.definition,
    getAllTodos.options
  )

  return isQueryLoading(rest) ? (
    <Spinner size="xxlarge" className="u-flex u-flex-justify-center" />
  ) : (
    <>
      <TodoList todos={data} />
      <TodoAdd />
    </>
  )
}

export default TodoWrapper
