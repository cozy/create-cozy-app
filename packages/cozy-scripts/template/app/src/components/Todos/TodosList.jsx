import React from 'react'

import TodoRemoveButton from './TodoRemoveButton'

export const TodosList = props => {
  const { todos } = props
  if (!todos || !todos.length) return null
  return (
    <div>
      <h2>Todos list</h2>
      <ul className="todos-list">
        {todos.map(todo => (
          <li key={todo._id} className="todo-item">
            <span>{todo.name}</span>
            <TodoRemoveButton todo={todo} />
          </li>
        ))}
      </ul>
    </div>
  )
}

export default TodosList
