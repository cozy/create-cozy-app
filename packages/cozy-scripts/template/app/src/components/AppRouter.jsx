import React from 'react'
import {
  Route,
  Navigate,
  RouterProvider,
  createHashRouter,
  createRoutesFromElements
} from 'react-router-dom'

import TodoWrapper from 'src/components/Todos/TodoWrapper'
import Hello1 from 'src/components/HelloViews/Hello1'
import Hello2 from 'src/components/HelloViews/Hello2'
import AppLayout from 'src/components/AppLayout'

const AppRouter = () => {
  const routes = (
    <Route path="/" element={<AppLayout />}>
      <Route path="todos" element={<TodoWrapper />} />
      <Route path="viewhello1" element={<Hello1 />} />
      <Route path="viewhello2" element={<Hello2 />} />
      <Route path="*" element={<Navigate to="/todos" replace />} />
    </Route>
  )
  const router = createHashRouter(createRoutesFromElements(routes))

  return <RouterProvider router={router} />
}

export default AppRouter
