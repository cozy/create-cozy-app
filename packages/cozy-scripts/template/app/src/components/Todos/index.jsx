import React from 'react'

import { useQuery, isQueryLoading } from 'cozy-client'
import Spinner from 'cozy-ui/transpiled/react/Spinner'

import { getAllTodos } from '../../utils/queries'
import TodoAdd from './TodoAdd/TodoAdd'
import TodosList from './TodoList/TodosList'

export const Todos = () => {
  const { data, ...rest } = useQuery(
    getAllTodos.definition,
    getAllTodos.options
  )

  return (
    <div className="todos">
      {isQueryLoading(rest) ? (
        <Spinner size="xxlarge" middle />
      ) : (
        <div>
          <TodosList todos={data} />
          <TodoAdd />
        </div>
      )}
    </div>
  )
}

export default Todos
