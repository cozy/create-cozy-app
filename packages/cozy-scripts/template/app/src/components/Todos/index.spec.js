'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

// we import the not connected component version for testing
import { Todos } from 'components/Todos'

// mock data for testing
import mockTodos from './_mocksTodos.json'

const getProps = (todos = [], fetchStatus = 'loaded') => ({
  todos: {
    data: todos,
    fetchStatus
  }
})

describe('Todos component:', () => {
  it('should be rendered correctly without todos', () => {
    const component = shallow(<Todos {...getProps()} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly with todos', () => {
    const component = shallow(<Todos {...getProps(mockTodos)} />).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should render a spinner if loading', () => {
    const component = shallow(
      <Todos {...getProps(mockTodos, 'loading')} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
