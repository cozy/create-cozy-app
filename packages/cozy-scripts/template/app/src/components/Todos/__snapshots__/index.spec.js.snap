// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos component: should be rendered correctly with todos 1`] = `
<div
  className="todos"
>
  <div>
    <TodosList
      todos={
        Array [
          Object {
            "_id": 1234,
            "name": "Todo 1",
          },
          Object {
            "_id": 5678,
            "name": "Todo 2",
          },
          Object {
            "_id": 9101,
            "name": "Todo 3",
          },
        ]
      }
    />
    <withClient(TodoAdd) />
  </div>
</div>
`;

exports[`Todos component: should be rendered correctly without todos 1`] = `
<div
  className="todos"
>
  <div>
    <TodosList
      todos={Array []}
    />
    <withClient(TodoAdd) />
  </div>
</div>
`;

exports[`Todos component: should render a spinner if loading 1`] = `
<div
  className="todos"
>
  <withI18n(Spinner)
    middle={true}
    size="xxlarge"
  />
</div>
`;
